core/types/transaction.go

```diff
@@ -310,25 +298,142 @@ func Fee(gasPrice *big.Int, gasLimit uint64, gatewayFee *big.Int) *big.Int {
 	return gasFee.Add(gasFee, gatewayFee)
 }
 
+// Data returns the input data of the transaction.
+func (tx *Transaction) Data() []byte { return tx.inner.data() }
+
+// AccessList returns the access list of the transaction.
+func (tx *Transaction) AccessList() AccessList { return tx.inner.accessList() }
+
+// Gas returns the gas limit of the transaction.
+func (tx *Transaction) Gas() uint64 { return tx.inner.gas() }
+
+// GasPrice returns the gas price of the transaction.
+func (tx *Transaction) GasPrice() *big.Int { return new(big.Int).Set(tx.inner.gasPrice()) }
+
+// GasTipCap returns the gasTipCap per gas of the transaction.
+func (tx *Transaction) GasTipCap() *big.Int { return new(big.Int).Set(tx.inner.gasTipCap()) }
+
+// GasFeeCap returns the fee cap per gas of the transaction.
+func (tx *Transaction) GasFeeCap() *big.Int { return new(big.Int).Set(tx.inner.gasFeeCap()) }
+
+// Value returns the ether amount of the transaction.
+func (tx *Transaction) Value() *big.Int { return new(big.Int).Set(tx.inner.value()) }
+
+// Nonce returns the sender account nonce of the transaction.
+func (tx *Transaction) Nonce() uint64 { return tx.inner.nonce() }
+
 // To returns the recipient address of the transaction.
-// It returns nil if the transaction is a contract creation.
+// For contract-creation transactions, To returns nil.
 func (tx *Transaction) To() *common.Address {
-	if tx.data.Recipient == nil {
+	// Copy the pointed-to address.
+	ito := tx.inner.to()
+	if ito == nil {
 		return nil
 	}
-	to := *tx.data.Recipient
-	return &to
+	cpy := *ito
+	return &cpy
+}
+
+// Cost returns value + gasprice * gaslimit + gatewayfee.
+func (tx *Transaction) Cost() *big.Int {
+	total := new(big.Int).Mul(tx.GasPrice(), new(big.Int).SetUint64(tx.Gas()))
+	total.Add(total, tx.Value())
+	if gatewayFee := tx.GatewayFee(); gatewayFee != nil {
+		total.Add(total, gatewayFee)
+	}
+	return total
 }
 
-// Hash hashes the RLP encoding of tx.
-// It uniquely identifies the transaction.
+// RawSignatureValues returns the V, R, S signature values of the transaction.
+// The return values should not be modified by the caller.
+func (tx *Transaction) RawSignatureValues() (v, r, s *big.Int) {
+	return tx.inner.rawSignatureValues()
+}
+
+// GasFeeCapCmp compares the fee cap of two transactions.
+func (tx *Transaction) GasFeeCapCmp(other *Transaction) int {
```

Comment:
I see there are 2 usages of this method. One that first checks that `other` has same feeCurrency, another that i'm unsure. We should check that.

Still, even if current usage are ok, we won't know for future usages. And i see that as quite dangerous, we do a future upstream merge which uses this method and everything breaks. 

Not sure what's the best approach here. One could be to rename the method to something that makes it very clear that `other` must have same currency. Another is to add a second argument (currency comparator); to do the proper comparison

Reply:
I checked all usages and it's properly converting before usage. Given how often we have to handle different tx currencies & the several different usages, it's probably worth thinking up a coherent design for how to handle this.

Reply:
Even if it's being used properly, i think the name and behaviour is missguiding, and can lead to future errors. So either change the name or fix the behaviour